specification/taf-spec.md

```diff
@@ -0,0 +1,915 @@
+- Changed: This signifies that the update was successful and detected changes.
+- Unchanged: This means the update was successful, but no changes were detected.
+- Failed: This event is triggered if an error occurred while running the updater.
+- Completed: Similar to a 'finally' block in programming, this event occurs regardless of the update's
+success or failure.
+
+For instance, if an update is successful and changes are detected, TAF will trigger the `changed`,
+`succeeded`, and `completed` events in sequence.
+
+In TAF, scripts are associated with specific events by being placed in designated folders within the
+targets/scripts directory of the authentication repository. This means that all script files are also
+TUF target files, and any modifications to them require signing the targets metadata. Each folder
+within this directory is named after the event it corresponds to and can contain multiple scripts.
+These scripts are executed in alphabetical order when their associated event is triggered. Below is an
+outline of the structure for the scripts folder:
+
+```markdown
+/scripts
+    /repo
+      /succeeded - every time a repo is successfully pulled
+      /changed
+      /unchanged
+      /failed - every time a repo is not successfully pulled
+      /completed  - like finally (called in both cases)
+    /update
+      /succeeded - once after all authentication's repositories have been successfully pulled
+      /changed
+      /unchanged
+      /failed - if one repository failed
+      /completed  - like finally (called in both cases)
+```
+
+In TAF's `targets/scripts` directory, there are `update` and `repo` folders that correspond to
+two levels of event triggers. The `repo` folder contains scripts triggered after updating a
+single authentication repository. In contrast, the `update` folder contains scripts that are
+executed once the entire update process for all repositories is complete.
+
+In situations where a TAF authentication repository references other authentication
+repositories, scripts in the repo folder will be executed multiple times. This execution occurs
+for each update of the referenced authentication repositories, as well as for the top-level
+authentication repository itself. On the other hand, scripts in the update folder are triggered
+only once, at the very end of the entire update process.
+
+In TAF, each script is expected to output JSON containing two types of data: persistent and
+transient. The persistent data is automatically saved to a file named `persistent.json` after
+every script execution. This file is then passed on to subsequent scripts. On the other hand,
+transient data, while also passed to the next script, is not stored permanently.
+In addition to handling persistent and transient data, these scripts also receive detailed
+information about the repositories involved — this includes both the authentication repository
+and its target repositories.
+
```

Comment:
To avoid repetition here and in updater docs, would it make sense to move this part to the full updater docs? And then we can simplify this section by just saying: for detailed explanation on the scripts, please refer to the Updater documentation and lifecycle hooks [here](https://github.com/openlawlibrary/taf/blob/a35dda05f14d817b6be680bbfabe228793b7fd8a/docs/updater/update_process.md?plain=1#L36)